### PR TITLE
Differentiate between instantiate and call methods.

### PR DESCRIPTION
**Pull request by shagunsodhani**
## Motivation

Differentiate between `instantiate` and `call` methods.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

Added tests

## Related Issues and PRs

https://github.com/facebookresearch/hydra/issues/731

### CODE REVIEW

**Comment by shagunsodhani** [comment]
@omry I could use some feedback on the PR. The current status is as follows:

1. Fails on lint tests.
2. Using `type` to denote the class will likely be a problem eventually (given its a Python keyword and used in Python files when using Structured Configs). 
3. Need to update the documentation. Will do that once point 2 is fixed.

**Comment by omry** [comment]
I will review soon, hopefully tomorrow.
in the mean time, can you give a proper title to the PR?

File: hydra/_internal/utils.py [hunk 1]
```diff
@@ -504,19 +504,36 @@ def _get_kwargs(config: Union[ObjectConf, DictConfig], **kwargs: Any) -> Any:
     return final_kwargs
 
 
+def _has_field(config: Union[ObjectConf, DictConfig], field: str) -> bool:
```
Comment (omry) [inline hydra/_internal/utils.py:507]: It's hard to review when you mix your changes with refactoring like moving this function as is.
Try to separate cleanup to a different commit labeled as such to make the review easier.

File: hydra/_internal/utils.py [hunk 2]
```diff
@@ -525,32 +542,60 @@ def _getcls(field: str) -> str:
...
         return _getcls(field="cls")
-    elif _has_field(field="class"):
+    elif _has_field(config=config, field="class"):
         return _getcls(field="class")
     else:
-        raise ValueError("Input config does not have a target field")
+        raise ValueError("Input config does not have a type field")
+
+
+def _get_callable_name(config: Union[ObjectConf, DictConfig]) -> str:
```
Comment (omry) [inline hydra/_internal/utils.py:564]: this function is a copy of the other one with some minor changes. try to reduce this duplication by passing additional parameters. to one function.

File: hydra/_internal/utils.py (hunk 1 above)
Comment (shagunsodhani) [reply hydra/_internal/utils.py:507]: My bad. Will be more careful next time.

File: hydra/_internal/utils.py (hunk 2 above)
Comment (shagunsodhani) [reply hydra/_internal/utils.py:564]: Okay!

**Comment by shagunsodhani** [comment]
> Nice.
> 
> I am thinking to add another mode that officially supports either type or callable as the same field.
> could still be target.
> 
> The advantage is that it simplifies the logic if the code does not know or care if it's a callable or a type.
> 
> we can make it explicit:
> `instantiate_or_call(cfg)`
> Where the it accepts without complains either type, callable or target.

Sounds good.

> Also, you didn't update the documentation and you didn't create a news fragment.

Yeah this is pending on alternative name of `type`. I am proposing to use `cls_name` or `cls_type` (or its modifications with `class`).

**Comment by omry** [comment]
let's go with type.
we can always deprecate it if we really get into trouble.

**Comment by omry** [comment]
Take a look at [this](https://github.com/anthonytec2/pytorch-lightning/pull/2/files/17450c54837b535757f5d89ba2ba63f116278a83..64714cbef1d8d6eb377c84d85ad207fa228b8582#diff-3598bc6c2f51dfaaf178b169ea9b5af4R33-R35).

Any thoughts about making instantiate(None) to return None to simplify that?

**Comment by omry** [comment]
cc @anthonytec2

**Comment by shagunsodhani** [comment]
> this

Yes! This sounds quite useful. Should I include it in the next PR ?

**Comment by omry** [comment]
or this one. you still need to make some changes here I think.

**Comment by shagunsodhani** [comment]
> or this one. you still need to make some changes here I think.

Right my bad :)

**Comment by shagunsodhani** [comment]
@omry I want to confirm if I should be updating all the plugins to use `type` or `callable` in place of `target`? Using `target` is supported.

Todo:

1. Take another pass with the documentation and examples (might have missed some)
2. Add support for handling null value
3. Update documentation and add news fragment.

File: hydra/_internal/utils.py [hunk 3]
```diff
@@ -522,86 +522,67 @@ def _has_field(config: Union[ObjectConf, DictConfig], field: str) -> bool:
...
+    """
+
+    if should_get_cls:
+        fn_name = "instantiate"
+        key_name = "type"
+        key_name_to_depriciate = "callable"
+    else:
+        fn_name = "call"
+        key_name = "callable"
+        key_name_to_depriciate = "type"
```
Comment (omry) [inline hydra/_internal/utils.py:547]: deprecate

File: hydra/_internal/utils.py [hunk 4]
```diff
@@ -522,86 +522,67 @@ def _has_field(config: Union[ObjectConf, DictConfig], field: str) -> bool:
...
-        return _getcls(field="class")
-    else:
-        raise ValueError("Input config does not have a type field")
 
+def _get_cls_or_callable_name(
+    config: Union[ObjectConf, DictConfig], should_get_cls: bool
+) -> str:
+    """Since the logic for getting class or callable name is very similar,
+    we use this wrapper function to avoid code-duplication. if `should_get_cls` is set to True,
+    the function returns the name of class, else it returns the name of callable.
```
Comment (omry) [inline hydra/_internal/utils.py:537]: no need for the comment to justify the function.
just describe what it does.

File: hydra/types.py [hunk 5]
```diff
@@ -14,6 +14,8 @@
 class ObjectConf(Dict[str, Any]):
     # class, class method or function name
     target: str = MISSING
+    callable: str = MISSING
+    type: str = MISSING
```
Comment (omry) [inline hydra/types.py:18]: I don't like this.
let's chat tomorrow.

File: website/docs/patterns/objects.md [hunk 6]
```diff
@@ -4,19 +4,30 @@ title: Creating objects and calling functions
...
     :param config: An ObjectConf or DictConfig describing what to call and what params to use
     :param args: optional positional parameters pass-through
     :param kwargs: optional named parameters pass-through
-    :return: the return value from the specified class or method
+    :return: the return value from the specified method
     """
 ```
 
-### ObjectConf definition
+### ObjectConf definition for hydra.utils.instantiate
```
Comment (omry) [inline website/docs/patterns/objects.md:30]: ObjectConf here does not line up with the abomination I don't like above.

**Comment by shagunsodhani** [comment]
> Take a look at [this](https://github.com/anthonytec2/pytorch-lightning/pull/2/files/17450c54837b535757f5d89ba2ba63f116278a83..64714cbef1d8d6eb377c84d85ad207fa228b8582#diff-3598bc6c2f51dfaaf178b169ea9b5af4R33-R35).
> 
> Any thoughts about making instantiate(None) to return None to simplify that?
> 
> `config.yaml`:
> 
> ```yaml
> foo: null
> ```
> 
> ```python
> cfg = OmegaConf.create("config.yaml")
> assert instantiate(cfg.foo) == None
> ```
> 
> `Structured Configs`:
> 
> ```python
> @dataclass
> class Config:
>   foo: Optional[ObjectConf] = None
> 
> cfg = OmegaConf.structured(Config)
> assert instantiate(cfg.foo) == None
> ```

https://github.com/facebookresearch/hydra/pull/737/commits/fea610e6807d7a62aa2b0c5c0ed2cfb4fd61a5eb adds support for this - for both `call` and `instantiate`. I do not feel strongly about supporting both `call` and `instantiate` and we may decided to support just `instantiate`.

**Comment by shagunsodhani** [comment]
Closing in light of recent discussions (https://github.com/facebookresearch/hydra/issues/731#issuecomment-656426570)
